File: backend/cxx/src/env.cxx
#include "node.h"
#include "env.h"
#include <iostream>
#include <cassert>

void VentureEnvironment::addBinding(VentureSymbol * vsym, Node * node) 
{ 
  frame.insert({vsym->sym,node}); 
  vsyms.push_back(vsym);
}

void VentureEnvironment::destroySymbols()
{
  for (VentureSymbol * vsym : vsyms) { delete vsym; }
}


Node * VentureEnvironment::findSymbol(VentureSymbol * vsym)
{
  return findSymbol(vsym->sym);
}

Node * VentureEnvironment::findSymbol(const string & sym)
{
  if (frame.count(sym))
  { 
    assert(frame[sym]);
    assert(frame[sym]->isValid());
    return frame[sym]; 
  }
  else if (outerEnv == nullptr)
  { 
    cout << "Cannot find symbol: " << sym << endl;
    assert(outerEnv);
    return nullptr;
  }
  else 
  {
    return outerEnv->findSymbol(sym);
  }
}

File: backend/jventure/env.jl
abstract Environment
require("value.jl")

type EmptyEnvironment <: Environment
end

type ExtendedEnvironment <: Environment
  frame::Dict{Symbol,Node}
  outerEnv::Environment
end

function ExtendedEnvironment(env::ExtendedEnvironment,syms::Array{VentureValue},nodes::Array{Node})
  return ExtendedEnvironment((Symbol=>Node)[sym => node for (sym,node) in zip(syms,nodes)],env)
end

findSymbol(env::ExtendedEnvironment,sym::Symbol) = haskey(env.frame,sym) ? env.frame[sym] : findSymbol(env.outerEnv,sym)
findSymbol(env::EmptyEnvironment,sym::Symbol) = error("Cannot find: " * string(sym))

function addBinding!(env::ExtendedEnvironment,sym::Symbol,node::Node)
  env.frame[sym] = node
end
